MaraBot/Commands/WeeklyCommandModule.cs

```diff
@@ -7,24 +8,42 @@
 namespace MaraBot.Commands
 {
     using Core;
+    using Messages;
     
     public class WeeklyCommandModule : BaseCommandModule
     {
-        public IWeeklyConfig Config { private get; set; }
-        
-        private const string k_ValidCommandEmoji = ":white_check_mark:";
-        private const string k_InvalidCommandEmoji = ":no_entry_sign:"; 
+        public Weekly Weekly { private get; set; }
+        public IReadOnlyDictionary<string, Preset> Presets { private get; set; }
         
         [Command("weekly")]
         [Cooldown(2, 900, CooldownBucketType.Channel)]
         [RequireGuild]
         public async Task Execute(CommandContext ctx)
         {
-            // TODO...
-            // 1. Randomly select a preset among list of weekly presets and generate a race.
-            // 2. In a more advanced state, write weekly seed to server log and repost the same
-            // race for the current week.
-            throw new NotImplementedException();
+            var weekNumber = RandomUtils.GetWeekNumber();
+
+            // Generate a new weekly seed
+            if (Weekly.WeekNumber != weekNumber)
+            {
+                Weekly = Weekly.Generate(Presets);
+                WeeklyIO.StoreWeekly(Weekly);
+            }
+
+            if (!Presets.ContainsKey(Weekly.PresetName))
```

Comment:
Don't we already know that the presets are valid? I'd suggest to only check for invalid weekly generation by using `Weekly.WeekNumber < 0`, and if so notify the user that there are no available presets, instead of that some preset is invalid.
This is especially important because an invalid weekly returns "" as the preset name, which makes the error look really weird.

Reply:
The presets may not necessarily contain the weekly preset if:
1- The weekly was generated and saved to json.
2- The bot stops executing and we change the presets somehow.
3- We relaunch the bot and weekly preset is not in the list anymore.

I really only added this check for safety in cases something went wrong. We could probably have a better message to the user though. like "Contact your friendly neighborhood developers" ;)

Also, we could add way more logs in the bot.
https://dsharpplus.github.io/articles/beyond_basics/logging/default.html
This kind of error should appear in the logs.

Otherwise, I agree with your comment, we should probably test for `Weekly.Invalid` and display a message/log in case the weekly is invalid :)